migration/idempotent/001-base.sql

```diff
@@ -3384,73 +3384,181 @@ $$
 LANGUAGE PLPGSQL STABLE;
 GRANT EXECUTE ON FUNCTION _prom_catalog.get_metrics_that_need_compression() TO prom_maintenance;
 
+DO $block$
+BEGIN
+    IF NOT _prom_catalog.is_timescaledb_installed() THEN
+        RETURN;
+    END IF;
+
+    CREATE OR REPLACE VIEW _prom_catalog.metric_chunks_that_need_to_be_compressed AS
+    SELECT
+        m.id AS metric_id,
+        m.metric_name,
+        m.hypertable_id,
+        m.table_name,
+        c.chunks_to_compress
+    FROM
+    (
+        SELECT
+            m.id,
+            m.metric_name,
+            m.table_name,
+            h.id AS hypertable_id
+        FROM _prom_catalog.metric m
+        INNER JOIN _timescaledb_catalog.hypertable h ON (h.schema_name OPERATOR(pg_catalog.=) 'prom_data' AND h.table_name OPERATOR(pg_catalog.=) m.table_name)
+        WHERE m.is_view IS FALSE
+        AND (m.delay_compression_until IS NULL OR m.delay_compression_until OPERATOR(pg_catalog.<) now())
+        AND h.compression_state OPERATOR(pg_catalog.=) 1 -- compression enabled
+        ORDER BY random() -- random metric order
+    ) m
+    INNER JOIN LATERAL -- lateral to force looking up chunks 1 metric at a time in nested loop
+    (
+        SELECT -- array of objects in order from oldest to newest
+            jsonb_agg
+            (
+                jsonb_build_object
+                (
+                    'chunk_id', x.chunk_id,
+                    'schema_name', x.schema_name,
+                    'table_name', x.table_name,
+                    'range_end', x.range_end,
+                    'chunk_num', x.chunk_num
+                )
+                ORDER BY x.chunk_num DESC -- oldest chunks first
+            )
+            FILTER
+            (
+                WHERE x.chunk_num OPERATOR(pg_catalog.>) 1 -- don't compress the most recent chunk
+                AND x.range_end OPERATOR(pg_catalog.<=) now() - interval '1 hour' -- don't compress chunks that aren't at least 1 hour old
```

Comment:
hmm if this was a function and not a view the 1 hour interval could be a parameter. We might need this feature

Reply:
As a view, I can use the view's type to declare variable in the execute_compression_policy.
As far as i know, there is no way to declare variables of a type referring to the return type of a function.
I can use record for a variable type, but a `record[]` is not allowed. Thus, I have to use a temp table to accumulate the metrics that we cannot lock without waiting.

Reply:
I guess I **could use a variable of type jsonb... that feels kinda gross though. WDYT?